frontend/javascripts/oxalis/model/sagas/quick_select_saga.ts

```diff
@@ -0,0 +1,596 @@
+  }
+
+  const inputData = normalizeToUint8(colorLayer, inputDataRaw, layerConfiguration);
+  const inputNdUvw = ndarray(inputData, size, stride).transpose(firstDim, secondDim, thirdDim);
+
+  const centerUV = take2(V3.floor(V3.scale(inputNdUvw.shape as Vector3, 0.5)));
+  // Two fields are computed (one for the dark segment scenario, and one for the light segment)
+  const darkThresholdField = getThresholdField(inputNdUvw, centerUV, "dark");
+  const lightThresholdField = getThresholdField(inputNdUvw, centerUV, "light");
+
+  // Copy unthresholded fields so that finetuning is possible
+  // without recomputing them. Only necessary if showPreview is true.
+  const unthresholdedDarkCopy =
+    quickSelectConfig.showPreview && copyNdArray(Uint8Array, darkThresholdField);
+  const unthresholdedLightCopy =
+    quickSelectConfig.showPreview && copyNdArray(Uint8Array, lightThresholdField);
+
+  const { initialDetectDarkSegment, darkMaxEffectiveThresh, lightMinEffectiveThresh } =
+    determineThresholds(darkThresholdField, lightThresholdField, inputNdUvw);
+
+  // thresholdField initially stores the threshold values (uint8), but is
+  // later processed to a binary mask.
+  let thresholdField;
+  if (initialDetectDarkSegment) {
+    thresholdField = darkThresholdField;
+    // In numpy-style this would be:
+    // thresholdField[:] = thresholdField[:] < darkMaxEffectiveThresh
+    ops.ltseq(thresholdField, darkMaxEffectiveThresh);
+    yield* put(
+      updateUserSettingAction("quickSelect", {
+        ...quickSelectConfig,
+        segmentMode: "dark",
+        threshold: darkMaxEffectiveThresh,
+      }),
+    );
+  } else {
+    thresholdField = lightThresholdField;
+    // In numpy-style this would be:
+    // thresholdField[:] = thresholdField[:] > darkMaxEffectiveThresh
+    ops.gtseq(thresholdField, lightMinEffectiveThresh);
+    yield* put(
+      updateUserSettingAction("quickSelect", {
+        ...quickSelectConfig,
+        segmentMode: "light",
+        threshold: lightMinEffectiveThresh,
+      }),
+    );
+  }
+
+  processBinaryMaskInPlaceAndAttach(thresholdField, quickSelectConfig, rectangleGeometry);
+
+  const overwriteMode = yield* select(
+    (state: OxalisState) => state.userConfiguration.overwriteMode,
+  );
+
+  if (!quickSelectConfig.showPreview) {
+    yield* finalizeQuickSelect(
+      rectangleGeometry,
+      volumeTracing,
+      activeViewport,
+      labeledResolution,
+      boundingBoxMag1,
+      thirdDim,
+      size,
+      firstDim,
+      secondDim,
+      inputNdUvw,
+      thresholdField,
+      overwriteMode,
+      labeledZoomStep,
+    );
+    return;
+  }
+
+  // Start an iterative feedback loop when preview mode is active.
+  while (true) {
+    const { finetuneAction, cancelQuickSelectAction, escape, enter, confirm } = (yield* race({
+      finetuneAction: take("FINE_TUNE_QUICK_SELECT"),
+      cancelQuickSelectAction: take("CANCEL_QUICK_SELECT"),
```

Comment:
hmm, I see that it's a bit weird to have both, but I created them because I had different use cases in mind:

- "cancel/confirm quick select" is for places where we know that exactly these actions are needed
- "Enter" / "Escape" (not "cancel"/"confirm" as you wrote) are generic actions which are only dispatched if the user hits the keys "enter" or "escape". depending on the context, these actions could do something else. I created these because I didn't want to map "enter" to "confirm quick select" (this felt too specific). for now, nobody else listens to enter/escape, so it's not really necessary, but it felt nicer still.

what do you think?

Reply:
Yes that makes perfect sense and I thought so too. My comment was merely directed at the inconsistency in variable naming between `cancelQuickSelectAction` and `confirm` :)

Reply:
ahh, I see :) makes sense & changed it!